sql/opt/exec/execbuilder: TestExecBuild failed

sql/opt/exec/execbuilder.TestExecBuild [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3405002&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3405002&tab=artifacts#/) on master @ [44c95054cfe0108df872d2cda7952fc92259426b](https://github.com/cockroachdb/cockroach/commits/44c95054cfe0108df872d2cda7952fc92259426b):


```
=== RUN   TestExecBuild
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExecBuild537227465
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestExecBuild
    logic.go:3424: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestExecBuild537227465
--- FAIL: TestExecBuild (140.15s)
=== RUN   TestExecBuild/5node/scan_parallel
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7  
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +testdata/scan_parallel:87: EXPLAIN (VEC) SELECT * FROM data WHERE a IN (0, 2, 4, 6, 8)
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +expected:
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +    │
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +    └ Node 1
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +      └ *colfetcher.ColBatchScan
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +    
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +but found (query options: "") :
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +    │
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +    └ Node 1
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +      └ *colexec.ParallelUnorderedSynchronizer
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +        ├ *colfetcher.ColBatchScan
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +        ├ *colfetcher.ColBatchScan
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +        ├ *colfetcher.ColBatchScan
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +        ├ *colfetcher.ColBatchScan
E210904 05:16:37.916542 467747 sql/logictest/logic.go:3632  [-] 7 +        └ *colfetcher.ColBatchScan
    logic.go:2311: 
         
        testdata/scan_parallel:87: EXPLAIN (VEC) SELECT * FROM data WHERE a IN (0, 2, 4, 6, 8)
        expected:
            │
            └ Node 1
              └ *colfetcher.ColBatchScan
            
        but found (query options: "") :
            │
            └ Node 1
              └ *colexec.ParallelUnorderedSynchronizer
                ├ *colfetcher.ColBatchScan
                ├ *colfetcher.ColBatchScan
                ├ *colfetcher.ColBatchScan
                ├ *colfetcher.ColBatchScan
                └ *colfetcher.ColBatchScan
--- done: testdata/scan_parallel with config 5node: 11 tests, 1 failures
E210904 05:16:37.916892 467747 sql/logictest/logic.go:3657  [-] 8  testdata/scan_parallel:94: too many errors encountered, skipping the rest of the input
    logic.go:3192: 
        testdata/scan_parallel:94: error while processing
    logic.go:3192: testdata/scan_parallel:94: too many errors encountered, skipping the rest of the input
        --- FAIL: TestExecBuild/5node/scan_parallel (2.95s)
=== RUN   TestExecBuild/5node
    --- FAIL: TestExecBuild/5node (85.63s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestExecBuild PKG=./pkg/sql/opt/exec/execbuilder TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-json
</p>
</details>
/cc @cockroachdb/sql-queries RaduBerinde
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecBuild.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/opt/exec/execbuilder/testdata/scan_parallel
+++ b/pkg/sql/opt/exec/execbuilder/testdata/scan_parallel
@@ -84,7 +84,9 @@ vectorized: true
   table: data@primary
   spans: [/0 - /0] [/2 - /2] [/4 - /4] [/6 - /6] … (1 more)
 
-query T
+# We allow to retry because the concurrency limit on the semaphore might not be
+# updated right away.
+query T retry
 EXPLAIN (VEC) SELECT * FROM data WHERE a IN (0, 2, 4, 6, 8)
 ----
 │

